### ISSUE TITLE
Remove Reliance on vbscript

### ISSUE THREAD

**Issue by agentoranger** [LesFerch/WinSetView#133]
Notably, vbscript is on the way out, All the while, I am sure you are very busy.  I would like to request that the reliance on vbscript is removed.  On my current systems WinSetView is no longer functional.  Thank you for your time and I appreciate the use I have enjoyed from WinSetView.

![image](https://github.com/user-attachments/assets/fc45364f-bbc5-4750-95e4-b568a0513d4a)

**Comment by LesFerch** [comment]
I'm currently working on converting the 2000 lines of VBScript code to JScript and the longer term project is a full rewrite in C#. I hope to have the JScript conversion done in the next few weeks, but it might be a few months.

VBScript is to be uninstalled by default starting sometime in 2027. Therefore, I guess these are work computers and your IT department has decided to remove VBScript from them. Is that correct?

**Comment by agentoranger** [comment]
That's correct, I removed vbscript from my own system, this happens to be a Windows 24H2 install, so at very least, I know my problem is self created.  I'm not great with vbscript, thought I might be able to help.  Any interest in contribution?

**Comment by LesFerch** [comment]
Thanks, but the code really isn't set up for multiple contributors. It's a one-person project for now, although I've had a bit of help with the language files. I'm reasonably confident I'll get a JScript version done by year end. I'll post back here when I've got something ready to test.

**Comment by idnovic** [comment]
Hi! Also selfmade problem here. 
I am going to install vbscript temporarily. 

Just wanted to suggest this
https://www.vbsedit.com/faq/convertexe.asp

It may be possible to convert the vbscript to an exe in the meantime.

**Comment by LesFerch** [comment]
WinSetView is already a VbsEdit HTML application. The "convert to exe" is an illusion. Tools such as VbsEdit, Bat2Exe, PS2Exe, Py2Exe, WinRAR, etc. wrap an Exe shell around a script.

In the case of HTML applications, VbsEdit provides two modes. The basic mode is a simple self extracting archive that extracts an HTA file and then runs it via MSHTA. The advanced mode (used for WinSetView) provides an Exe shell that uses a webbrowser object (i.e. MSHTML) to run the HTML/CSS/VBScript code. The advanced mode replaces MSHTA and nothing is extracted to files (it all happens in memory) and it provides some improved functionality over MSHTA. The advanced mode requires Visual Studio to compile the Exe, but it's still a wrapper around a script.

The WinSetView VBScript code has to be carefully converted to [JScript](https://en.wikipedia.org/wiki/JScript) in order to run on machines where VBScript has been removed. ChatGPT helps, but it's non-trivial for so much code.

**Comment by idnovic** [comment]
Thank you for the explanation!

Why did you choose JScript? 
I have no experience with it. 

Why not powershell? 
C# as choice does make sense to me.

**Comment by LesFerch** [comment]
[JScript](https://en.wikipedia.org/wiki/JScript) is the only language that is a direct drop in replacement for VBScript in an HTML application. The HTML and CSS code will not have to be touched. The interface will remain identical and the extensive logic that dynamically builds much of the interface will remain the same. Any other language choice means starting from scratch.

When and if I rebuild the interface from scratch, C# makes the most sense. Here's why:

1. PowerShell does not directly provide GUI features. PowerShell programs that have a GUI rely on a third-party library or they use inline C# code. Either way, it's a hack that doesn't allow for a nice development cycle in an IDE. I don't see any point in using C# code in PowerShell when I can write directly in C#.

2. I am now comfortable with C# and Visual Studio, so I can build a C# app with a GUI interface in a proper development environment which helps a lot when the amount of code gets into the hundreds or thousands of lines.

3. A C# exe, using WinForms or WPF XAML on .Net 4.8, will run on Windows 7 through Windows 11 without installing any extra software. And the end-result exe is tiny. I like that.

**Comment by agentoranger** [comment]
> PS2Exe

Fun fact, PS2EXE uses a custom PowerShell interpreter and converts the resultant code into C#

**Comment by LesFerch** [comment]
I just tested PS2Exe to be sure and, as expected, it's a C# exe wrapper around the original script. It may replace PowerShell.exe, but not the DLLs that actually interpret the script. It's very similar to the way a C++ or C# exe wrapper with a webbrowser object replaces MSHTA.exe.

**Comment by LesFerch** [comment]
I've completed the conversion from VBScript to JScript.

It's available now as a beta. You can find it [here](https://github.com/LesFerch/WinSetView/releases/tag/3.0.1).

**Comment by LesFerch** [comment]
@agentoranger Have you tested the JScript version?

**Comment by Candlemass** [comment]
> @agentoranger Have you tested the JScript version?

@LesFerch

I've tested he beta with negative results. Still running into a similar problem being experienced by agentoranger.

Are you sure this is fully converted to C#?

I'm still getting an HTML dependency error requring the install of internet explorer and the corresponding updates in order to run the program which I and many others will not do nor should it be necessary.

https://i.postimg.cc/dt4D68QW/Error.jpg

Can we just get a simple batch file to convert all folder views to "general items" and "medium icons" system wide regardless of the file types currently within them? This all seems rather unnecesarily complicated just to change the values of a couple registry entries.

**Comment by Candlemass** [comment]
Nevermind, I reread what you posted and I guess the C# version is still yet to be completed. Will wait for that or painstakingly change every folder over tediously by hand that won't switch with the "apply to all folders" button in the folder view options menu.

**Comment by LesFerch** [comment]
@Candlemass Your issue is not VBScript vs JScript. Your issue is that you're missing MSHTML 11. That's not a lot to ask. It was released in 2013. Presumably, you're running Windows 7, but are missing that update. MSHTML 11 comes with IE 11. Just install the IE 11 update and you'll be good to go. Once you're updated, you can remove IE if you like. WinSetView only needs MSHTML 11.

Here's the link for the 64 bit version:

https://www.microsoft.com/en-us/download/details.aspx?id=41628

**Comment by Candlemass** [comment]
> @Candlemass Your issue is not VBScript vs JScript. Your issue is that you're missing MSHTML 11. That's not a lot to ask. It was released in 2013. Presumably, you're running Windows 7, but are missing that update. MSHTML 11 comes with IE 11. Just install the IE 11 update and you'll be good to go. Once you're updated, you can remove IE if you like. WinSetView only needs MSHTML 11.
> 
> Here's the link for the 64 bit version:
> 
> https://www.microsoft.com/en-us/download/details.aspx?id=41628

Yes, I already acknowledged this with my second post. Sorry but installing the cancer that is EI is not an option. Will wait for the C# version or skip altogether. I could throw together a batch script that completes this in probably the same amount of time it would take to properly uninstall ei 11 by manually tracking down all the garbage left behind in the registry and orphan files after it's "uninstalled".

**Comment by LesFerch** [comment]
Didn't your Windows 7 come with IE 8 pre-installed and then you removed it? Seems to me it would be simple enough to upgrade and remove IE again if you did it before. Maybe you even scripted that?

Anyhow, you could edit Win7.ini to your preferences and then run `WinSetView.ps1 win7.ini`.

**Comment by Candlemass** [comment]
I came to check on the status of this project only to find that you deleted my comments and your replies to them, why? I now have no idea what your last response was and neither will anyone else with a similar issue seeking a solution...

Anyways, I have downloaded the beta release you linked to. Still getting the exact same error whining about updates to EI...

**Comment by LesFerch** [comment]
@Candlemass I deleted your comments because this issue is about VBScript.

Your issue Is that WinSetView requires MSHTML 11 and your Windows 7 build only has MSHTML 8 (or nothing at all if you ripped that out). It's a different issue, so it should be opened as a separate issue.

The documentation has been updated with the specific [System Requirements](https://github.com/LesFerch/WinSetView/blob/main/System-Requirements.md)

**Comment by Candlemass** [comment]
> [JScript](https://en.wikipedia.org/wiki/JScript) is the only language that is a direct drop in replacement for VBScript in an HTML application. The HTML and CSS code will not have to be touched. The interface will remain identical and the extensive logic that dynamically builds much of the interface will remain the same. Any other language choice means starting from scratch.
> 
> **If** and when I rebuild the interface from scratch, C# makes the most sense. Here's why:
> 
> 1. PowerShell does not directly provide GUI features. PowerShell programs that have a GUI rely on a third-party library or they use inline C# code. Either way, it's a hack that doesn't allow for a nice development cycle in an IDE. I don't see any point in using C# code in PowerShell when I can write directly in C#.
> 2. I am now comfortable with C# and Visual Studio, so I can build a C# app with a GUI interface in a proper development environment which helps a lot when the amount of code gets into the hundreds or thousands of lines.
> 3. A C# exe, using WinForms or WPF XAML on .Net 4.8, will run on Windows 7 through Windows 11 without installing any extra software. And the end-result exe is tiny. I like that.

@LesFerch

The thread title is 100% relevant to the problem hence why I posted here.

Your choice of scripting language is the issue and because of this is reliant upon an antiquated browser engine that hasn't been used for it's intended purpose since 2015...

Now you mentioned a rewrite to a C# program which would solve the problem. As would a conversion to a batch script.

I also see within the folder powershell components. This could also potentially work.

Is there an update on any of this rather then creating a dependency on the cancer that is internet explorer and the deeply embedded install that takes hours to properly uninstall by hand through the registry since simply uninstalling via "add/remove programs" leaves hundreds of components behind?

**Comment by LesFerch** [comment]
The tool works just fine on the vast majority of machines out there. With the change to JScript it will be very close to 100%. How many users do you think are running a Windows 7 machine that was never updated to MSHTML 11?

Hundreds of hours of work have gone into this tool. It's no small task to rewrite it from the ground up. GUIs are hard work. Really hard work. I may rewrite it in C# some day (as I've already mentioned) but it will likely take me a year or more (I do have other things to do).

While related, this really should be a separate issue. I'm going to open a new issue and copy and paste our conversation there.

If you want to use the tool, install MSHTML 11. You can uninstall IE and probably can even delete the entire Internet Explorer folder. That's all I can offer as solution for now. It's a free tool. I do what I can, but I can't make everyone happy.

**Comment by LesFerch** [comment]
@agentoranger Have you tested the JScript version? I haven't heard from anyone about it yet, so your feedback would be most appreciated. Thanks!

**Comment by LesFerch** [comment]
Version 3.1.1 has now been released. All VBScript code has been replaced with JScript code.
